File: internal/app/service/utils/parse.go
package utils

import (
	"fmt"
	"log/slog"
	"mime/multipart"
	"strings"

	"github.com/labstack/echo/v4"
)

type (
	FileType struct {
		Type     string
		MimeType string
		Size     int64
	}
)

func ValidateFileHeader(fileHeader *multipart.FileHeader, fileTypes []FileType, keyName string) (err error) {
	filenameSplit := strings.Split(fileHeader.Filename, ".")
	fileType := strings.ToLower(filenameSplit[len(filenameSplit)-1])
	mimeType := fileHeader.Header.Get("Content-Type")
	for _, allowedType := range fileTypes {
		if fileType == allowedType.Type {
			if allowedType.Size != 0 && fileHeader.Size > allowedType.Size<<20 {
				err = fmt.Errorf("Error:Field validation for %s. File is too large", keyName)
			}
			return
		}
		if mimeType == allowedType.MimeType {
			if allowedType.Size != 0 && fileHeader.Size > allowedType.Size<<20 {
				err = fmt.Errorf("Error:Field validation for %s. File is too large", keyName)
			}
			return
		}
	}
	err = fmt.Errorf("Error:Field validation for %s. Invalid filetype", keyName)
	return
}

func ParseMultipartForm(c echo.Context, fileTypes []FileType) (files map[string][]*multipart.FileHeader, values map[string][]string, err error) {
	ctx := c.Request().Context()
	err = c.Request().ParseMultipartForm(0)
	if err != nil {
		slog.ErrorContext(ctx, "failed to parsemultipartformfile", "error", err)
		err = fmt.Errorf("Error:Field validation for parse multipart form")
		return
	}
	files = c.Request().MultipartForm.File
	values = c.Request().MultipartForm.Value
	if len(files) < 1 && len(values) < 1 && len(fileTypes) > 6 && len(values) > 6 {
		err = fmt.Errorf("Error:Field validation for parse multipart form")
		slog.ErrorContext(ctx, "failed to parsemultipartformfile", "error", err)
		return
	}
	if len(fileTypes) > 0 {
		for keyName, valuess := range files {
			for _, fileHeader := range valuess {
				err = ValidateFileHeader(fileHeader, fileTypes, keyName)
				if err != nil {
					err = fmt.Errorf("%s, %s", keyName, err)
					slog.ErrorContext(ctx, "failed to parsemultipartformfile", "error", err)
					return
				}
			}
		}
	}
	return
}

File: src/utils/parse.ts
import { Request } from "express";

export interface FileType {
  type: string; // contoh: "png", "jpg"
  mimeType: string; // contoh: "image/png"
  size: number; // ukuran maksimal (MB)
}

/**
 * Validate file header (type, mime, size)
 */
export function validateFileHeader(
  file: Express.Multer.File,
  fileTypes: FileType[],
  keyName: string
): Error | null {
  const filenameSplit = file.originalname.split(".");
  const fileType = filenameSplit[filenameSplit.length - 1].toLowerCase();
  const mimeType = file.mimetype;

  for (const allowedType of fileTypes) {
    if (fileType === allowedType.type) {
      if (allowedType.size !== 0 && file.size > allowedType.size * (1 << 20)) {
        return new Error(
          `Error: Field validation for ${keyName}. File is too large`
        );
      }
      return null;
    }

    if (mimeType === allowedType.mimeType) {
      if (allowedType.size !== 0 && file.size > allowedType.size * (1 << 20)) {
        return new Error(
          `Error: Field validation for ${keyName}. File is too large`
        );
      }
      return null;
    }
  }

  return new Error(`Error: Field validation for ${keyName}. Invalid filetype`);
}

/**
 * Parse multipart form: return files & values
 * (gunakan middleware multer di route)
 */
export function parseMultipartForm(
  req: Request,
  fileTypes: FileType[]
): {
  files: { [fieldname: string]: Express.Multer.File[] };
  values: { [key: string]: string[] };
} {
  if (!req.files && !req.body) {
    throw new Error("Error: Field validation for parse multipart form");
  }

  const files: { [fieldname: string]: Express.Multer.File[] } = {};
  const values: { [key: string]: string[] } = {};

  // ambil semua file
  if (req.files) {
    if (Array.isArray(req.files)) {
      files["files"] = req.files;
    } else {
      for (const [key, val] of Object.entries(req.files)) {
        files[key] = val as Express.Multer.File[];
      }
    }
  }

  // ambil semua values (body text fields)
  for (const [key, val] of Object.entries(req.body)) {
    values[key] = Array.isArray(val) ? (val as string[]) : [val as string];
  }

  // validasi file types
  if (fileTypes.length > 0) {
    for (const [key, fileList] of Object.entries(files)) {
      for (const file of fileList) {
        const err = validateFileHeader(file, fileTypes, key);
        if (err) {
          throw new Error(`${key}, ${err.message}`);
        }
      }
    }
  }

  return { files, values };
}
